File: 3298-count-substrings-that-can-be-rearranged-to-contain-a-string-ii/3298-count-substrings-that-can-be-rearranged-to-contain-a-string-ii.java
class Solution {
    boolean match (int[] cou1,int[] cou2){
        for (int i=0;i<26;i++){
            if (cou1[i]<cou2[i]){
                return false;
            }
        }
        return true;
    }
    public long validSubstringCount(String word1, String word2) {
        int[] cou=new int[26];
        int[] cou2=new int[26];
        for (int i=0;i<word2.length();i++){
            cou2[word2.charAt(i)-97]++;
        }
        int n=word1.length();
        int st=0;long ret=0L;
        for (int i=0;i<n;i++){
            cou[word1.charAt(i)-97]++;
            if (match(cou,cou2)){
                while (match(cou,cou2)){
                    cou[word1.charAt(st)-97]--;
                    st++;
                }
                st--;cou[word1.charAt(st)-97]++;
                ret=ret+st+1;
            }
        }
        return ret;   
    }
}

File: 3298-count-substrings-that-can-be-rearranged-to-contain-a-string-ii/3298-count-substrings-that-can-be-rearranged-to-contain-a-string-ii.py
class Solution:
    def validSubstringCount(self, word1: str, word2: str) -> int:
        N = len(word1)
        need = [0]*26
        for c in word2:
            need[ord(c)-ord('a')] +=1
        have = [0]*26
        letters_needed = [i for i in range(26) if need[i]>0]
        left = 0
        total = 0
        for right in range(N):
            c = ord(word1[right])-ord('a')
            have[c] +=1
            while all(have[i]>=need[i] for i in letters_needed):
                have[ord(word1[left])-ord('a')] -=1
                left +=1
            total += left
        return total        